figwheel-main/doc/figwheel-main-options.md
# Figwheel Main Configuration Options

The following options can be supplied to `figwheel.main` via the `figwheel-main.edn` file.

# Commonly used options (in order of importance)

## :watch-dirs

A list of ClojureScript source directories to be watched and compiled on change.

    :watch-dirs ["cljs-src"]

## :css-dirs

A list of CSS source directories to be watched and reloaded into the browser.

    :css-dirs ["resource/public/css"]

## :ring-handler

A symbol or string indicating a ring-handler to embed in the
figwheel.repl server. This aids in quickly getting a dev server up and
running. If the figwheel server doesn't meet your needs you can simply
start your own server, the figwheel.client will still be able to
connect to its websocket endpoint.
Default: none

    :ring-handler my-project.server/handler

## :ring-server-options

All the options to forward to the `ring-jetty-adapter/run-jetty` function
which figwheel.main uses to run its ring server.

All the available options are documented here:
https://github.com/ring-clojure/ring/blob/master/ring-jetty-adapter/src/ring/adapter/jetty.clj#L127

This will normally be used to set the `:port` and `:host` of the server.

Most uses of these options are considered advanced if you find
yourself using many of these options you problably need to run your
own server outside of figwheel.main.

## :rebel-readline

By default Figwheel engauges a Rebel readline editor when it starts
the ClojureScript Repl in the terminal that it is launched in.

This will only work if you have `com.bhauman/rebel-readline-cljs` in
your dependencies.

More about Rebel readline:
https://github.com/bhauman/rebel-readline

Default: true

    :rebel-readline false

## :pprint-config

When `:pprint-config` is set to true. The `figwheel.main` will print the
computed config information and will terminate the process. Useful for
understanding what figwheel.main adds to your configuration before it
compiles your build.

Default: false

    :pprint-config true

## :open-file-command

A path to an executable shell script that will be passed a file and
line information for a particular compilation error or warning.

A script like this would work
ie. in  ~/bin/myfile-opener

    #! /bin/sh
    emacsclient -n +$2:$3 $1

The add this script in your config:

    :open-file-command "myfile-opener"

But thats not the best example because Figwheel handles `emacsclient`
as a special case so as long as `emacsclient` is on the shell path you can
simply do:

    :open-file-command "emacsclient"

and Figwheel will call emacsclient with the correct args.

## :figwheel-core

Wether to include the figwheel.core library in the build. This
 enables hot reloading and client notification of compile time errors.
 Default: true

    :figwheel-core false

## :hot-reload-cljs

Whether or not figwheel.core should hot reload compiled
ClojureScript. Only has meaning when :figwheel is true.
Default: true

    :hot-reload-cljs false

## :reload-dependents

Whether or not figwheel.core should reload reload the namespaces
that `depend` on the changed namespaces in addition to the changed
namespaces themselves. Only has meaning when :figwheel is true.
Default:true

    :reload-dependents false

## :connect-url

The url that the figwheel repl client will use to connect back to
the server.

This url is actually a template that will be filled in.  For example
the default `:connect-url` is:

    "ws://[[config-hostname]]:[[server-port]]/figwheel-connect"

The available template variables are:

For the server side:

    [[config-hostname]]  the host supplied in :ring-server-options > :host or "localhost"
    [[server-hostname]]  the java.InetAddress localhost name - "Bruces-MacBook-Pro.local" on my machine
    [[server-ip]]        the java.InetAddress localhost ip interface - normally 192.168.x.x
    [[server-port]]      the port supplied in :ring-server-options > :port or the default port 9500

On the client side:

    [[client-hostname]]  the js/location.hostname on the client
    [[client-port]]      the js/location.port on the client

If the url starts with a Websocket scheme "ws://" a websocket
connection will be established. If the url starts with an http scheme
"http" an http long polling connection will be established.

## :open-url

Either a boolean value `false` or a string that indicates the url
that the figwheel repl will open in the browser after the source code
has been compiled. A `false` value will disable this behavior.

The string value is actually a template that can provide optional
template variables. For example the default `:open-url` is:

    "http://[[server-hostname]]:[[server-port]]"

The available template variables are:

For the server side:

    [[server-hostname]]  the host supplied in :ring-server-options > :host or "localhost"
    [[server-port]]      the port supplied in :ring-server-options > :port or the default port 9500

## :reload-clj-files

Figwheel naively reloads `clj` and `cljc` files on the `:source-paths`.
It doesn't reload clj dependent files like tools.namspace.

Figwheel does note if there is a macro in the changed `clj` or `cljc` file
and then marks any cljs namespaces that depend on the `clj` file for
recompilation and then notifies the figwheel client that these
namespaces have changed.

If you want to disable this behavior:

    :reload-clj-files false

Or you can specify which suffixes will cause the reloading

    :reload-clj-files #{:clj :cljc}

## :log-file

The name of a file to redirect the figwheel.main logging to. This
will only take effect when a REPL has been started.

    :log-file "figwheel-main.log"

## :log-level

The level to set figwheel.main java.util.logger to.
Can be one of: `:error` `:info` `:debug` `:trace` `:all` `:off`

    :log-level :error

## :client-log-level

The log level to set the client side goog.log.Logger to for
figwheel.repl and figwheel.core. Can be one of:
`:severe` `:warning` `:info` `:config` `:fine` `:finer` `:finest`

    :client-log-level :warning

## :log-syntax-error-style

figwheel.main logging prints out compile time syntax errors which
includes displaying the erroneous code.
Setting `:log-syntax-error-style` to `:concise` will cause the logging to
not display the erroneous code.
Available options: `:verbose`, `:concise`
Default: `:verbose`

    :log-syntax-error-style :concise

## :load-warninged-code

If there are warnings in your code emitted from the compiler, figwheel
does not refresh. If you would like Figwheel to load code even if
there are warnings generated set this to true.
Default: false

    :load-warninged-code true

## :ansi-color-output

Figwheel makes an effort to provide colorful text output. If you need
to prevent ANSI color codes in figwheel output set `:ansi-color-output`
to false.  Default: true

    :ansi-color-output false

## :validate-config

Whether to validate the figwheel-main.edn and build config (i.e.".cljs.edn") files.
Default: true

    :validate-config false

## :validate-cli

Whether to validate the figwheel-main command line options
Default: true

    :validate-cli false

## :target-dir

A String that specifies the target directory component of the path
where figwheel.main outputs compiled ClojureScript

The default `:output-dir` is composed of:

    [[:target-dir]]/public/cljs-out/[[build-id]]

The default `:output-to` is composed of:

    [[:target-dir]]/public/cljs-out/[[build-id]]-main.js

If you are using the default figwheel.repl server to serve compiled
assets, it is very important that the :target-dir be on the classpath.

The default value of `:target-dir` is "target"

    :target-dir "cljs-target"

## :launch-node

A boolean that indicates wether you want figwheel to automatically
launch Node. Defaults to true.

## :inspect-node

A boolean that indicates wether you want figwheel to enable remote
inspection by adding "--inspect" when it launches Node.
Defaults to true.

## :node-command

A String indicating the Node.js executable to launch Node with.
Defaults to "node"

## :cljs-devtools

A boolean that indicates wether to include binaryage/devtools into
the your clojurescript build. Defaults to true when the target is a
browser and the :optimizations level is :none, otherwise it is false.

    :cljs-devtools false

# Rarely used options

## :client-print-to

The `figwheel.repl` client can direct printed (via pr) output to the
repl and or the console. `:client-print-to` is a list of where you
want print output directed. The output choices are `:console` and `:repl`
Default: [:console :repl]

    :client-print-to [:console]

## :ring-stack

The fighweel server has a notion of a `:ring-stack`. The
`:ring-stack` is a composition of basic ring-middleware (think
sessions) to wrap around a supplied `:ring-handler`.

The default `:ring-stack` is a slightly modified
`ring.middleware.defaults/wrap-defaults`

## :ring-stack-options

The fighweel.repl server has a notion of a `:ring-stack`. The
`:ring-stack` is a composition of basic ring-middleware to wrap around
a supplied `:ring-handler`.

The default `:ring-stack` is a slightly modified
ring.middleware.defaults/wrap-defaults.

`:ring-stack-options` are the options that figwheel.repl supplies to
`ring.middleware.defaults/wrap-defaults`.

The default options are slightly modified from `ring.middleware.defaults/site-defaults`:

```
{:params
 {:urlencoded true, :multipart true, :nested true, :keywordize true},
 :cookies true,
 :session
 {:flash true, :cookie-attrs {:http-only true, :same-site :strict}},
 :static {:resources "public"},
 :responses {:content-types true, :default-charset "utf-8"},
 :figwheel.server.ring/dev
 {:figwheel.server.ring/fix-index-mime-type true,
  :figwheel.server.ring/resource-root-index true,
  :figwheel.server.ring/wrap-no-cache true,
  :ring.middleware.not-modified/wrap-not-modified true,
  :co.deps.ring-etag-middleware/wrap-file-etag true,
  :ring.middleware.cors/wrap-cors true,
  :ring.middleware.stacktrace/wrap-stacktrace true}}
```

You can override these options by suppling your own to `:ring-stack-options`

If these options are changed significantly don't be suprised if the
figwheel stops behaving correctly :)

## :wait-time-ms

The number of milliseconds to wait before issuing reloads. Set this
higher to wait longer for changes. This is the interval from when the first
file change occurs until we finally issue a reload event.

Default: 50

    :wait-time-ms 50

## :mode

The `:mode` indicates the behavior that occurs after a compile.
Options: `:repl` `:serve` or `:build-once`

* `:repl` indicates that repl sill be started
* `:serve` indicates that a server will be started
* `:build-once` indicates that a compile will not be follwed by any action

This is mainly intended for use when you are launching figwheel.main from a script.

Normally defaults to `:repl`

## :broadcast-reload

Figwheel broadcasts hot reloads to all clients that have connected
since the figwheel process has started. Set `:broadcast-reload` to
`false` if you want to only send hot-reloads to the client where the
REPL eval occurs.
Default: true

    :broadast-reload false

## :broadcast

In the past figwheel would broadcast REPL evaluations to all
connected clients and then print the first result received in the
REPL. Setting `:broadcast` to `true` will give you back this legacy
behavior. Default: false

    :broadcast true

## :repl-eval-timeout

The time (in milliseconds) it takes for the repl to timeout.
Evaluating any given expression in cljs can take some time.
The repl is configured to throw a timeout exception as to not hang forever.

This config option will determine how long the repl waits for the result of an eval
before throwing.

Default: 8000

    :repl-eval-timeout 10000 ;;waits for 10 seconds instead of 8

## :hawk-options

If you need to watch files with polling instead of FS events. This can
be useful for certain docker environments.

    :hawk-options {:watcher :polling}

sidecar/resources/conf-fig-docs/BuildOptionsMap.txt
:BuildOptionsMap

A Map of options that specifies a ClojureScript 'build'

  :BuildOptionsMap [:CreateExampleMap :id :source-paths :figwheel :compiler]

:id

A Keyword, String or Symbol that identifies this build.

  :id "dev"

:source-paths

A vector of paths to your cljs source files. These paths should be
relative from the root of the project to the root the namespace.
For example, if you have an src/example/core.cljs file that contains a
example.core namespace, the source path to this file is "src"

  :source-paths ["src"]

:notify-command

If a :notify-command is specified, it will be called when compilation
succeeds or fails, and a textual description of what happened will be
appended as the last argument to the command. If a more complex
command needs to be constructed, the recommendation is to write a
small shell script wrapper.
Default: nil (disabled)

  :notify-command ["growlnotify" "-m"]

:initial-notify-command

This option works like :notify-command, but specifies a command to be called
after initial compilation. :notify-command is called only for compilations
triggered in reaction to file changes.
Default: nil (disabled)

  :initial-notify-command ["growlnotify" "-m"]


CHANGES.md
## 0.5.16 Caching improvements & no trampoline on windows

Daniel Compton did some great work solving the caching issues.

There is also a very confusing problem with lein trampoline on Windows
see issue #682. So I turned off the auto-trampoline feature for
Windows which allowed the for the elegent use of Rebel readline by
just calling `lein figwheel`.

Windows users can still elect to use trampoline with figwheel and get
the benefits of rebel-readline but the will have to call lein
trampoline figwheel and this should be done with the knowledge that
they may experience a classpath corruption issue.

* worker target support PR #659
* important caching tweaks for serving CLJS assets PR #667
* :npm-deps false support PR #678
* fix to prefer and detect new cider/piggieback

## 0.5.15 Readline

A big thanks goes out to ClojuristsTogether.org and everyone who
contributed for helping to make this work possible!

This release includes the biggest change that Figwheel has had in
quite a while. After 2 months of work I'm proud to release a version
of Figwheel that has Clojure readline support. Even if you always use
your editor REPL like me, you may want to try the REPL that `lein
figwheel` launches. You may find that its a handy tool to have around.

You can learn more about rebel-readline at
[https://github.com/bhauman/rebel-readline](https://github.com/bhauman/rebel-readline).

Don't forget this gets rid of the need for `rlwrap`.

You can disable the Rebel readline by adding `:readline false` to your
`:figwheel` config.

This release requires that you use
`[org.clojure/clojurescript "1.9.946"]` at least.

This release also, adds automatic support for projects that use
[Code Splitting](https://clojurescript.org/guides/code-splitting). You
should now be able to use code splitting and have Figwheel "just work".

Other changes:

* Figwheel is now using cljs `:preloads` for figwheel client injection

## 0.5.14 Faster loading for complex dependency trees

Figwheel inherited a topo sort algorithm from the CLJS compiler. It
turned out that this graph sorting algorithm missed a very important
optimization, which is fixed in this release. If you have project that
has a lot of namespaces and have experienced slower load times when
you change a file that is deeper in your dependency chain, you should
notice a big improvement when using this release.

Added JavaScript environment hooks for tools like `re-natal` to
customize the websocket implementation and the script loading behavior.

* improved topo-sort algorithm
* CLJS compile option: the `:entries` key is no longer required in the
  `:modules` configs
* new `:repl-eval-timeout` allows you to increase or lower the
  Figwheel REPL eval timeout as a top level config setting
* added `goog.global.FIGWHEEL_WEBSOCKET_CLASS` to allow one to
  override or supply a websocket implementation for the figwheel
  client
* added `goog.global.FIGWHEEL_IMPORT_SCRIPT` to allow one to override
  or supply a function to the figwheel client, that is responsible for
  loading a namespace into the JavaScript runtime.
* added two new namespaces that can be supplied the ClojureScript
  compiler's `:preloads` option.
  
  The two namespaces are only different in that one set's up figwheel
  while the other both sets up and starts figwheel. These are
  currently meant to clean up how the figwheel client is injected into
  the build. The next release of Figwheel will do away with generating
  a small ClojureScript file to inject the Figwheel client into build.
  - `figwheel.connect` which will take a configuration from
    `:external-tooling` > `:figwheel/config` and supply a
    `figwheel.connect/start` function which contains the supplied
    config options. This function is exported so that it can be easily
    called from JavaScript.
  - `figwheel.preload` which simply calls the above `figwheel.connect/start`
    function

## 0.5.13 Small updates

* remove the use of a deprecated Google Closure library function 
  goog.net.jsloader.load
* add CLJS compile option `:process-shim` to validation code

## 0.5.12 Cache busting and more compiler option updates

There have been constant problems when reloading an application into
Chrome where even a hard reload insists on pulling items from the
cache. Adding `no-cache` to the response headers headers and returning
appropriate `not-modified` responses is a trade-off that will
hopefully address this issue without adversely affecting initial load
performance. Keep in mind that this will only apply to Figwheel's built-in
dev server. If this becomes a problem for some projects we can explore
adding a flag for this behavior.

* let :npm-deps use any named type for the keys 
* allow the server to serve symlinks outside of the project
* take more care when checking localStorage for persistent config
* add suspendable interface implementation to figwheel system component
* add :checked-arrays option
* update the catalog of :warning and :closure-warning options
* fixes uses of `aget`
* make configuration smiley paredit friendly

## 0.5.11 New cljs compiler options

* add CLJS compile option `:fn-invoke-direct` to validation schema
* add CLJS compile option `:source-map-asset-path` to validation schema
* add CLJS compile option `:load-test` to validation schema
* add CLJS compile option `:npm-deps` to validation schema
* only watch directories if they exist - PR #551 fixes Issue #451

I knew when I started validating compiler options that there was a
trade-off. On the one hand I can create really specific errors
(i.e. spelling errors) if I do a hard check for key membership, but
with the down side that as figwheel trails the compiler development
new options would couse the config to fail validiation.

I figured this would be a good trade off considering that:

A. Compiler options will be added at a slower and slower rate as time
goes on.

B. In general newer compiler options are geared towards more advanced

figwheel-main/src/figwheel/main/schema/config.clj
(ns figwheel.main.schema.config
  (:require
   [clojure.java.io :as io]
   [clojure.string :as string]
   [clojure.spec.alpha :as s]
   [clojure.set]
   [figwheel.main.util :as util]
   [figwheel.main.schema.core
    :as schema
    :refer [def-spec-meta non-blank-string? directory-exists?
            ensure-all-registered-keys-included]]
   [expound.alpha :as exp]
   [spell-spec.alpha :as spell]
   [spell-spec.expound]))

(s/def ::watch-dirs
  (s/coll-of
   (s/and non-blank-string?
          directory-exists?
          ::schema/has-cljs-source-files)))

#_(exp/expound ::watch-dirs ["/Users/bhauman/workspace/temp/figtest/ouchy"])

(def-spec-meta ::watch-dirs
  :doc
  "A list of ClojureScript source directories to be watched and compiled on change.

    :watch-dirs [\"cljs-src\"]"
  :group :common)

(s/def ::css-dirs (s/coll-of (s/and non-blank-string?
                                    directory-exists?)))
(def-spec-meta ::css-dirs
  :doc
  "A list of CSS source directories to be watched and reloaded into the browser.

    :css-dirs [\"resource/public/css\"]"
  :group :common)

(s/def ::ring-handler (s/or :non-blank-string non-blank-string?
                            :symbol ::schema/unquoted-symbol))
(def-spec-meta ::ring-handler
  :doc
  "A symbol or string indicating a ring-handler to embed in the
figwheel.repl server. This aids in quickly getting a dev server up and
running. If the figwheel server doesn't meet your needs you can simply
start your own server, the figwheel.client will still be able to
connect to its websocket endpoint.
Default: none

    :ring-handler my-project.server/handler"
  :group :common)

(s/def ::ring-server-options (s/keys :opt-un
                                     [::schema/integer-port
                                      ::schema/host]))

(def-spec-meta ::ring-server-options
  :doc
 "All the options to forward to the `ring-jetty-adapter/run-jetty` function
which figwheel.main uses to run its ring server.

All the available options are documented here:
https://github.com/ring-clojure/ring/blob/master/ring-jetty-adapter/src/ring/adapter/jetty.clj#L127

This will normally be used to set the `:port` and `:host` of the server.

Most uses of these options are considered advanced if you find
yourself using many of these options you problably need to run your
own server outside of figwheel.main."
  :group :common)

(s/def ::rebel-readline boolean?)
(def-spec-meta ::rebel-readline
  :doc
  "By default Figwheel engauges a Rebel readline editor when it starts
the ClojureScript Repl in the terminal that it is launched in.

This will only work if you have `com.bhauman/rebel-readline-cljs` in
your dependencies.

More about Rebel readline:
https://github.com/bhauman/rebel-readline

Default: true

    :rebel-readline false"
  :group :common)

(s/def ::pprint-config boolean?)
(def-spec-meta ::pprint-config
  :doc
  "When `:pprint-config` is set to true. The `figwheel.main` will print the
computed config information and will terminate the process. Useful for
understanding what figwheel.main adds to your configuration before it
compiles your build.

Default: false

    :pprint-config true"
  :group :common)

(s/def ::open-file-command non-blank-string?)
(def-spec-meta ::open-file-command
  :doc
  "A path to an executable shell script that will be passed a file and
line information for a particular compilation error or warning.

A script like this would work
ie. in  ~/bin/myfile-opener

    #! /bin/sh
    emacsclient -n +$2:$3 $1

The add this script in your config:

    :open-file-command \"myfile-opener\"

But thats not the best example because Figwheel handles `emacsclient`
as a special case so as long as `emacsclient` is on the shell path you can
simply do:

    :open-file-command \"emacsclient\"

and Figwheel will call emacsclient with the correct args."
  :group :common)

(s/def ::figwheel-core boolean?)
(def-spec-meta ::figwheel-core
  :doc
 "Wether to include the figwheel.core library in the build. This
 enables hot reloading and client notification of compile time errors.
 Default: true

    :figwheel-core false"
  :group :common)

(s/def ::hot-reload-cljs boolean?)
(def-spec-meta ::hot-reload-cljs
  :doc
 "Whether or not figwheel.core should hot reload compiled
ClojureScript. Only has meaning when :figwheel is true.
Default: true

    :hot-reload-cljs false"
  :group :common)

(s/def ::reload-dependents boolean?)
(def-spec-meta ::reload-dependents
  :doc
 "Whether or not figwheel.core should reload reload the namespaces
that `depend` on the changed namespaces in addition to the changed
namespaces themselves. Only has meaning when :figwheel is true.
Default:true

    :reload-dependents false"
  :group :common)

(s/def ::connect-url non-blank-string?)
(def-spec-meta ::connect-url
  :doc
 "The url that the figwheel repl client will use to connect back to
the server.

This url is actually a template that will be filled in.  For example
the default `:connect-url` is:

    \"ws://[[config-hostname]]:[[server-port]]/figwheel-connect\"

The available template variables are:

For the server side:

    [[config-hostname]]  the host supplied in :ring-server-options > :host or \"localhost\"
    [[server-hostname]]  the java.InetAddress localhost name - \"Bruces-MacBook-Pro.local\" on my machine
    [[server-ip]]        the java.InetAddress localhost ip interface - normally 192.168.x.x
    [[server-port]]      the port supplied in :ring-server-options > :port or the default port 9500

On the client side:

    [[client-hostname]]  the js/location.hostname on the client
    [[client-port]]      the js/location.port on the client

If the url starts with a Websocket scheme \"ws://\" a websocket
connection will be established. If the url starts with an http scheme
\"http\" an http long polling connection will be established."
  :group :common)

(s/def ::open-url (s/or :non-blank-string non-blank-string?
                        :false false?))
(def-spec-meta ::open-url
  :doc
 "Either a boolean value `false` or a string that indicates the url
that the figwheel repl will open in the browser after the source code
has been compiled. A `false` value will disable this behavior.

The string value is actually a template that can provide optional
template variables. For example the default `:open-url` is:

    \"http://[[server-hostname]]:[[server-port]]\"

The available template variables are:

For the server side:

    [[server-hostname]]  the host supplied in :ring-server-options > :host or \"localhost\"
    [[server-port]]      the port supplied in :ring-server-options > :port or the default port 9500"
  :group :common)

(s/def ::reload-clj-files (s/or :bool boolean?
                                :extension-coll (s/coll-of #{:clj :cljc})))
(def-spec-meta ::reload-clj-files
  :doc
 "Figwheel naively reloads `clj` and `cljc` files on the `:source-paths`.
It doesn't reload clj dependent files like tools.namspace.

Figwheel does note if there is a macro in the changed `clj` or `cljc` file
and then marks any cljs namespaces that depend on the `clj` file for
recompilation and then notifies the figwheel client that these
namespaces have changed.

If you want to disable this behavior:

    :reload-clj-files false

Or you can specify which suffixes will cause the reloading

    :reload-clj-files #{:clj :cljc}"
  :group :common)

(s/def ::log-file non-blank-string?)
(def-spec-meta ::log-file
  :doc
 "The name of a file to redirect the figwheel.main logging to. This
will only take effect when a REPL has been started.

    :log-file \"figwheel-main.log\""
  :group :common)

(s/def ::log-level #{:error :info :debug :trace :all :off})
(def-spec-meta ::log-level
  :doc
 "The level to set figwheel.main java.util.logger to.
Can be one of: `:error` `:info` `:debug` `:trace` `:all` `:off`

    :log-level :error"
  :group :common)

(s/def ::client-log-level  #{:severe :warning :info :config :fine :finer :finest})
(def-spec-meta ::client-log-level
  :doc
 "The log level to set the client side goog.log.Logger to for
figwheel.repl and figwheel.core. Can be one of:
`:severe` `:warning` `:info` `:config` `:fine` `:finer` `:finest`

    :client-log-level :warning"
  :group :common)

(s/def ::log-syntax-error-style #{:verbose :concise})
(def-spec-meta ::log-syntax-error-style
  :doc
 "figwheel.main logging prints out compile time syntax errors which
includes displaying the erroneous code.
Setting `:log-syntax-error-style` to `:concise` will cause the logging to
not display the erroneous code.
Available options: `:verbose`, `:concise`
Default: `:verbose`

    :log-syntax-error-style :concise"
  :group :common)

(s/def ::load-warninged-code boolean?)
(def-spec-meta ::load-warninged-code
  :doc
 "If there are warnings in your code emitted from the compiler, figwheel
does not refresh. If you would like Figwheel to load code even if
there are warnings generated set this to true.
Default: false

    :load-warninged-code true"
  :group :common)

(s/def ::ansi-color-output boolean?)
(def-spec-meta ::ansi-color-output
  :doc
   "Figwheel makes an effort to provide colorful text output. If you need
to prevent ANSI color codes in figwheel output set `:ansi-color-output`
to false.  Default: true

    :ansi-color-output false"
  :group :common)

(s/def ::validate-config boolean?)
(def-spec-meta ::validate-config
  :doc
 "Whether to validate the figwheel-main.edn and build config (i.e.\".cljs.edn\") files.
Default: true

    :validate-config false"
  :group :common)

(s/def ::validate-cli boolean?)
(def-spec-meta ::validate-cli
  :doc
 "Whether to validate the figwheel-main command line options
Default: true

    :validate-cli false"
  :group :common)

(s/def ::target-dir non-blank-string?)
(def-spec-meta ::target-dir
  :doc
 "A String that specifies the target directory component of the path
where figwheel.main outputs compiled ClojureScript

The default `:output-dir` is composed of:

    [[:target-dir]]/public/cljs-out/[[build-id]]

The default `:output-to` is composed of:

    [[:target-dir]]/public/cljs-out/[[build-id]]-main.js

If you are using the default figwheel.repl server to serve compiled
assets, it is very important that the :target-dir be on the classpath.

The default value of `:target-dir` is \"target\"

    :target-dir \"cljs-target\""
  :group :common)

(s/def ::launch-node boolean?)
(def-spec-meta ::launch-node
  :doc
  "A boolean that indicates wether you want figwheel to automatically
launch Node. Defaults to true."
  :group :common)

(s/def ::inspect-node boolean?)
(def-spec-meta ::inspect-node
  :doc
  "A boolean that indicates wether you want figwheel to enable remote
inspection by adding \"--inspect\" when it launches Node.
Defaults to true."
  :group :common)

(s/def ::node-command non-blank-string?)
(def-spec-meta ::node-command
  :doc
  "A String indicating the Node.js executable to launch Node with.
Defaults to \"node\""
  :group :common)

(s/def ::cljs-devtools boolean?)

(def-spec-meta ::cljs-devtools
  :doc
  "A boolean that indicates wether to include binaryage/devtools into
the your clojurescript build. Defaults to true when the target is a
browser and the :optimizations level is :none, otherwise it is false.

    :cljs-devtools false"
  :group :common)

;; -------------------------------XXXXXXXXXXXX

(s/def ::client-print-to (s/coll-of #{:console :repl}))
(def-spec-meta ::client-print-to
  :doc
 "The `figwheel.repl` client can direct printed (via pr) output to the
repl and or the console. `:client-print-to` is a list of where you
want print output directed. The output choices are `:console` and `:repl`
Default: [:console :repl]

    :client-print-to [:console]"
  :group :un-common)

(s/def ::ring-stack (s/or :non-blank-string non-blank-string?
                          :symbol ::schema/unquoted-symbol))

(def-spec-meta ::ring-stack
  :doc
 "The fighweel server has a notion of a `:ring-stack`. The
`:ring-stack` is a composition of basic ring-middleware (think
sessions) to wrap around a supplied `:ring-handler`.

The default `:ring-stack` is a slightly modified
`ring.middleware.defaults/wrap-defaults`"
  :group :un-common)

(s/def ::ring-stack-options map?)
(def-spec-meta ::ring-stack-options
  :doc
 (str "The fighweel.repl server has a notion of a `:ring-stack`. The
`:ring-stack` is a composition of basic ring-middleware to wrap around
a supplied `:ring-handler`.

The default `:ring-stack` is a slightly modified
ring.middleware.defaults/wrap-defaults.

`:ring-stack-options` are the options that figwheel.repl supplies to
`ring.middleware.defaults/wrap-defaults`.

The default options are slightly modified from `ring.middleware.defaults/site-defaults`:

```
" (when-let [opt (resolve 'figwheel.server.ring/default-options)]
    (with-out-str (clojure.pprint/pprint (deref opt))))
"```

You can override these options by suppling your own to `:ring-stack-options`

If these options are changed significantly don't be suprised if the
figwheel stops behaving correctly :)")
  :group :un-common)

(s/def ::wait-time-ms integer?)
(def-spec-meta ::wait-time-ms
  :doc
  "The number of milliseconds to wait before issuing reloads. Set this
higher to wait longer for changes. This is the interval from when the first
file change occurs until we finally issue a reload event.

Default: 50

    :wait-time-ms 50"
  :group :un-common)

(s/def ::mode #{:build-once :repl :serve})
(def-spec-meta ::mode
  :doc
 "The `:mode` indicates the behavior that occurs after a compile.
Options: `:repl` `:serve` or `:build-once`

* `:repl` indicates that repl sill be started
* `:serve` indicates that a server will be started
* `:build-once` indicates that a compile will not be follwed by any action

This is mainly intended for use when you are launching figwheel.main from a script.

Normally defaults to `:repl`"
:group :un-common)

(s/def ::broadcast-reload boolean?)
(def-spec-meta ::broadcast-reload
  :doc
  "Figwheel broadcasts hot reloads to all clients that have connected
since the figwheel process has started. Set `:broadcast-reload` to
`false` if you want to only send hot-reloads to the client where the
REPL eval occurs.
Default: true

    :broadast-reload false"

  :group :un-common)

(s/def ::broadcast boolean?)
(def-spec-meta ::broadcast
  :doc
  "In the past figwheel would broadcast REPL evaluations to all
connected clients and then print the first result received in the
REPL. Setting `:broadcast` to `true` will give you back this legacy
behavior. Default: false

    :broadcast true"
  :group :un-common)

(s/def ::repl-eval-timeout integer?)
(def-spec-meta ::repl-eval-timeout
  :doc
  "The time (in milliseconds) it takes for the repl to timeout.
Evaluating any given expression in cljs can take some time.
The repl is configured to throw a timeout exception as to not hang forever.

This config option will determine how long the repl waits for the result of an eval
before throwing.

Default: 8000

    :repl-eval-timeout 10000 ;;waits for 10 seconds instead of 8"
  :group :un-common)

(s/def ::hawk-options (s/map-of #{:watcher} #{:barbary :java :polling}))

(def-spec-meta ::hawk-options
  :doc
  "If you need to watch files with polling instead of FS events. This can
be useful for certain docker environments.

    :hawk-options {:watcher :polling}"
  :group :un-common)


(s/def ::edn
  (ensure-all-registered-keys-included
   #{::edn}
   (spell/strict-keys
    :opt-un
    [::watch-dirs
     ::css-dirs
     ::ring-handler
     ::ring-server-options
     ::rebel-readline
     ::pprint-config
     ::open-file-command
     ::figwheel-core
     ::hot-reload-cljs
     ::reload-dependents
     ::connect-url
     ::open-url
     ::reload-clj-files
     ::log-file
     ::log-level
     ::client-log-level
     ::log-syntax-error-style
     ::load-warninged-code
     ::ansi-color-output
     ::validate-config
     ::validate-cli
     ::target-dir

     ::launch-node
     ::inspect-node
     ::node-command
     ::cljs-devtools

     ::client-print-to
     ::ring-stack
     ::ring-stack-options
     ::wait-time-ms
     ::mode
     ::ring-server
     ::broadcast
     ::broadcast-reload
     ::repl-eval-timeout
     ::hawk-options
     ])))
